<h1 align=center> CUARTA PARTE DEL PROYECTO: MODELO DE RECOMENDACIÓN </h1>

El objetivo es armar un sistema de recomendación de películas para usuarios, donde dado un id de usuario y una película, nos diga si la recomienda o no para dicho usuario.

---

<h3 align=center> Recursos </h3>

In [60]:
import numpy as np
import pandas as pd

from surprise import Dataset, Reader, SVD
from surprise.model_selection import train_test_split

import matplotlib

---

<h3 align=center> Data </h3>

In [61]:
df = pd.read_csv('./data/data_ML.csv')
df.head()

id  userId  rating                  title
0  as680       1     1.0  the english civil war
1  as680     583     4.5  the english civil war
2  as680     765     5.0  the english civil war
3  as680    2116     3.0  the english civil war
4  as680    2143     3.0  the english civil war

Nota: El archivo data_ML.csv no fue subido a GitHub por superar los 300 megas. Antes de corres este módulo es necesatio volver al módulo de EDA (EDA.ipynb) y correr el último bloque: `df.to_csv('data/data_ML.csv', encoding='utf-8', index=False)`

In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10792583 entries, 0 to 10792582
Data columns (total 4 columns):
 #   Column  Dtype  
---  ------  -----  
 0   id      object 
 1   userId  int64  
 2   rating  float64
 3   title   object 
dtypes: float64(1), int64(1), object(2)
memory usage: 329.4+ MB


In [63]:
reader = Reader(rating_scale=(1, 5))
df_1 = df[['id', 'userId', 'rating']].sample(n=100000, random_state=0)
data = Dataset.load_from_df(df_1, reader)
# El método sólo permite utiizar 100k registros


In [64]:
df_1.tail(n=10)

id  userId  rating
3172845    hs583  123449     2.5
5891058   as2736    5993     2.0
909218    ns5276    6222     4.0
1883503    ns645  117215     2.5
10020113  ns7642   68854     4.5
7662561   ns8036  122347     3.0
4486860   as4297   60185     4.0
10242883  as3413  266061     3.0
6667825    hs807   15086     3.0
4601913   hs2201   54735     5.0

---

#### 1. El primer paso siempre es instanciar el modelo, en este caso SVD (_Singular Value Decomposition_)

In [65]:
model = SVD()

---

#### 2. Dividimos nuestros datos en datos de test y datos de entrenamiento

In [66]:
train_data, test_data = train_test_split(data, test_size=0.3)

---

#### 3. Entrenamiento del modelo utilizando el método `fit()`

In [67]:
model.fit(train_data)

---

#### 4. Aplicamos el modelo a los datos de test

In [68]:
predictions = model.test(test_data)

In [69]:
predictions[0:10]

[Prediction(uid='ns7874', iid=30950, r_ui=3.0, est=3.534721047106184, details={'was_impossible': False}),
 Prediction(uid='as3122', iid=67897, r_ui=2.0, est=3.1173395512340565, details={'was_impossible': False}),
 Prediction(uid='as3371', iid=70648, r_ui=3.0, est=3.185253281776118, details={'was_impossible': False}),
 Prediction(uid='hs866', iid=40291, r_ui=4.0, est=3.3159531316008315, details={'was_impossible': False}),
 Prediction(uid='as4890', iid=13777, r_ui=5.0, est=3.741023973737038, details={'was_impossible': False}),
 Prediction(uid='as6975', iid=17364, r_ui=3.0, est=3.6352253629773785, details={'was_impossible': False}),
 Prediction(uid='ns5170', iid=119556, r_ui=4.0, est=3.5313357142857145, details={'was_impossible': False}),
 Prediction(uid='ns872', iid=250949, r_ui=4.0, est=3.5795734470809375, details={'was_impossible': False}),
 Prediction(uid='as8881', iid=259284, r_ui=3.0, est=3.752192495296696, details={'was_impossible': False}),
 Prediction(uid='ns683', iid=16948, r_ui

---

<h3 align=center> Predicciones </h3>

Dado el ususario '5924' y la película 'ds206' la predicción que se obtiene es la sigueinte:

In [70]:
model.predict(5924, 'ds206')

Prediction(uid=5924, iid='ds206', r_ui=None, est=3.5313357142857145, details={'was_impossible': False})

Se estima que el usuario anterior puede llegar a evaluar a `ds206` en 3.5

---

<h3 align=center> Referencias </h3>

[1] https://surprise.readthedocs.io/en/stable/ </br>
[2] https://rubialesalberto.medium.com/singular-value-decomposition-svd-con-python-creando-un-sistema-de-recomendacion-640de1b8b002 </br>
[3] https://interactivechaos.com/es/wiki/descomposicion-en-valores-singulares-svd </br>

